rcl_action/src/rcl_action/graph.c

```diff
@@ -0,0 +1,216 @@
+  size_t j = 0u;
+  for (size_t i = 0u; i < num_names; ++i) {
+    const char * identifier_index = strstr(names[i], action_name_identifier);
+    if (identifier_index) {
+      const size_t action_name_len = strlen(names[i]) - suffix_len;
+      char * action_name = rcutils_strndup(names[i], action_name_len, *allocator);
+      if (!action_name) {
+        RCL_SET_ERROR_MSG("Failed to allocate memory for action name");
+        ret = RCL_RET_BAD_ALLOC;
+        break;
+      }
+
+      action_names_and_types->names.data[j] = action_name;
+
+      // Allocate storage for type list
+      rcutils_ret_t rcutils_ret = rcutils_string_array_init(
+        &action_names_and_types->types[j],
+        topic_names_and_types->types[i].size,
+        allocator);
+      if (RCUTILS_RET_OK != rcutils_ret) {
+        RCL_SET_ERROR_MSG(rcutils_get_error_string().str);
+        ret = RCL_RET_BAD_ALLOC;
+        break;
+      }
+
+      // Populate types list
+      for (size_t k = 0u; k < topic_names_and_types->types[i].size; ++k) {
+        char * type_name = topic_names_and_types->types[i].data[k];
+        size_t action_type_len = strlen(type_name);
+        // Trim type name suffix
+        const size_t type_suffix_len = strlen(action_type_identifier);
+        const char * type_identifier_index = strstr(type_name, action_type_identifier);
+        if (type_identifier_index) {
+          action_type_len = strlen(type_name) - type_suffix_len;
+        }
+        // Copy name to output struct
+        char * action_type_name = rcutils_strndup(type_name, action_type_len, *allocator);
+        if (!action_type_name) {
+          RCL_SET_ERROR_MSG("Failed to allocate memory for action type");
+          ret = RCL_RET_BAD_ALLOC;
+          break;
+        }
+        action_names_and_types->types[j].data[k] = action_type_name;
+      }
+      ++j;
+    }
+  }
+
+  // Cleanup if there is an error
+  if (RCL_RET_OK != ret) {
+    rcl_ret_t fini_ret = rcl_names_and_types_fini(action_names_and_types);
+    (void)fini_ret;  // Error already set
+  }
+
+  return ret;
+}
+
+rcl_ret_t
+rcl_action_get_client_names_and_types_by_node(
+  const rcl_node_t * node,
+  rcl_allocator_t * allocator,
+  const char * node_name,
+  const char * node_namespace,
+  rcl_names_and_types_t * action_names_and_types)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(action_names_and_types, RCL_RET_INVALID_ARGUMENT);
+
+  rcl_ret_t ret;
+  rcl_names_and_types_t topic_names_and_types = rcl_get_zero_initialized_names_and_types();
+  ret = rcl_get_subscriber_names_and_types_by_node(
+    node, allocator, false, node_name, node_namespace, &topic_names_and_types);
+  if (RCL_RET_OK != ret) {
+    return ret;
+  }
+
+  ret = _filter_action_names(
+    &topic_names_and_types,
+    allocator,
+    action_names_and_types);
+
+  rcl_ret_t nat_fini_ret = rcl_names_and_types_fini(&topic_names_and_types);
+  if (RCL_RET_OK != nat_fini_ret) {
+    return nat_fini_ret;
```

Comment:
Recommend `fini`'ing `action_names_and_types` here so the user doesn't have to do it when the function returns an error code, same below.

Reply:
Done 57e2280